src/repr/src/row.rs

```diff
@@ -533,9 +546,29 @@ fn push_datum<T: Bytes>(data: &mut T, datum: Datum) {
         }
         Datum::JsonNull => data.push(Tag::JsonNull as u8),
         Datum::Dummy => data.push(Tag::Dummy as u8),
-        Datum::APD(n) => {
+        Datum::APD(mut n) => {
+            // Pseudo-canonical representation of decimal values with
+            // insignificant zeroes trimmed. This compresses the number further
+            // than `Apd::trim` by removing all zeroes, and not only those in
+            // the fractional component.
+            apd::cx_datum().reduce(&mut n.0);
+            let (digits, exponent, bits, lsu) = n.0.to_raw_parts();
+            // SAFETY: `lsu` is a `&[u16]`, so each element can safely be
+            // transmuted into two `u8`s.
+            let (prefix, lsu, suffix) = unsafe { lsu.align_to::<u8>() };
+            // Each element of the LSU is 3-digits wide, so the raw LSU has
+            // essentially ceil(digits / 3) elements; the u8 aligned version
+            // should have twice that many.
+            assert!(lsu.len() == ((usize::try_from(digits).unwrap() + 2) / 3) * 2);
+            // There should be no unaligned elements in the prefix or suffix.
+            assert!(prefix.is_empty() && suffix.is_empty());
             data.push(Tag::APD as u8);
-            push_copy!(data, n, OrderedDecimal<Apd>);
+            // Digits cannot exceed 39
+            push_copy!(data, u8::try_from(digits).unwrap(), u8);
+            // Exponent cannot exceed +/-39
+            push_copy!(data, i8::try_from(exponent).unwrap(), i8);
+            data.push(bits);
+            push_untagged_bytes(data, &lsu);
```

Comment:
Since the length of LSU is entirely determined by `digits`, it’ll be better just to blat down the LSU bytes directly here. (`push_untagged_bytes` will unnecessarily write the bytes down again with an 8 byte length prefix, if I’m reading it right.)

Reply:
Implemented this suggestion